packages/dendron-next-server/components/graph.tsx

```diff
@@ -0,0 +1,156 @@
+import { createLogger } from "@dendronhq/common-frontend";
+import _ from "lodash";
+import React, { useEffect, useRef, useState } from "react";
+import cytoscape, { Core, ElementsDefinition, EventHandler } from "cytoscape";
+// @ts-ignore
+import euler from "cytoscape-euler";
+import { useRouter } from "next/router";
+import { useThemeSwitcher } from "react-css-theme-switcher";
+import { Space, Typography } from "antd";
+import Head from "next/head";
+
+const getCytoscapeStyle = (themes: any, theme: string | undefined) => `
+      node {
+        width: 5;
+        height: 5;
+        background-color: ${theme === themes.dark ? "#807B7B" : "#999393"};
+        border-color: ${theme === themes.dark ? "#807B7B" : "#999393"};
+        color: ${theme === themes.dark ? "#fff" : "#2F3438"};
+        label: data(label);
+        border-width: 1;
+        font-size: 10;
+        min-zoomed-font-size: 10;
+        font-weight: 400;
+      }
+    
+      edge {
+        width: 0.25;
+        line-color: ${theme === themes.dark ? "#807B7B" : "#999393"};
+        target-arrow-shape: none;
+        curve-style: haystack;
+      }
+    
+      :selected, .open {
+        background-color: ${theme === themes.dark ? "#36B73B" : "#27AC2C"};
+        border-color: ${theme === themes.dark ? "#36B73B" : "#27AC2C"};
+        color: ${theme === themes.dark ? "#36B73B" : "#27AC2C"};
+      }
+    `;
+
+export default function Graph({
+  elements,
+  type = "note",
+  onSelect,
+}: {
+  elements: ElementsDefinition | undefined;
+  onSelect: EventHandler;
+  type?: "note" | "schema";
+}) {
+  const router = useRouter();
+
+  const logger = createLogger("Graph");
+  const graphRef = useRef<HTMLDivElement>(null);
+  const { switcher, themes, currentTheme, status } = useThemeSwitcher();
+
+  const [cy, setCy] = useState<Core>();
+
+  useEffect(() => {
+    if (graphRef.current && elements) {
+      // Naive check to prevent full graph re-renders when selecting a node
+      if (cy && cy.elements("*").length > 5) return;
```

Comment:
can you elaborate on the `length > 5`?

Reply:
This line is designed to prevent the graph from re-rendering when selecting a node. I added this check due to the way the tendency of the schema graph to render twice - once with the single "hard-coded" root node, and once again once the schema data is available. I can see if there's a more elegant solution where the schema graph only gets rendered after the schema data is available

Reply:
makes sense. was confused about 5 vs 1 but see that you updated it to 1

Reply:
Yeah, that was just to have a buffer as I was working things out and I had forgotten to set it to a more understandable/accurate value